### ISSUE TITLE
Var is Beyond Slow

### ISSUE THREAD

**Issue by RobertBColton** [enigma-dev/enigma-dev#1617]
@rpjohnst said it before and now it's time for me to confirm it, ENIGMA's var is unbelievably slow and we need to improve its performance. I have several graphics benchmarks that really outperform GMSv1.4 when I switch them from var to int. The simplest example I have is a benchmark that draws 5000 sprites at random locations. Just using int in place of var for the location more than doubles the framerate from 297~ fps to 687~ fps. Using variant is almost as fast as the int also and definitely faster than GMSv1.4.

Forum post by TheExDeus when he originally raised the issue: https://enigma-dev.org/forums/index.php?topic=1911

Download Draw Sprite Benchmark: [draw_sprite_var_benchmark.zip](https://github.com/enigma-dev/enigma-dev/files/2973530/draw_sprite_var_benchmark.zip)

| Max FPS | Var | Variant | float | double | Int |
|---------|-----|-----|-----|-----|-----|
| GL1     | 🔻 284 | ✔️ 587 | ✔️ 610 | ✔️ 612 | ✔️ 603 |
| DX9     | 🔻 300 | ✔️ 661 | ✔️ 667 | ✔️ 674 | ✔️ 689 |
| GMSv1.4 | ✔️ 466 | ➖ N/A | ➖ N/A | ➖ N/A | ➖ N/A |

```cpp
// create event
room_speed = 99999;
set_synchronization(false);

// draw event
repeat (5000) {
    var xx, yy;
    xx = random(room_width);
    yy = random(room_height);
    draw_sprite(spr_0, 0, xx, yy);
}
room_caption = string(fps);
```

**Comment by time-killer-games** [comment]
Last I checked room_caption was taken out after Studio 1.3 was released.

**Comment by RobertBColton** [comment]
Change it to `window_set_caption(string(fps));` and remove the call to `set_synchronization(false);` which is also deprecated.

**Comment by time-killer-games** [comment]
Yeah i know, i was just making an observation.

**Comment by JoshDreamland** [comment]
I can investigate why var has become so slow this coming weekend. I'm swamped with family stuff right now, and I have work to contend with as well. Bug me Wednesday evening; I may have time, then.

**Comment by fundies** [comment]
just make jdi support c++17 and we can use std::variant

**Comment by RobertBColton** [comment]
The state of this issue has been improved with #1644, but we can go farther yet. As the benchmark I ran in that pull request shows, we can do more to optimize the set time of a var.

**Comment by RobertBColton** [comment]
Closing as resolved by #1676.
